src/main/cpp/native_glue.cc

```diff
@@ -19,140 +19,74 @@
-}
-
-void litecore::jni::UTF8CharToModifiedUTF8(const char *input, char *output) {
-    char c = input[0];
-    char c1 = input[1] & 0x3F;
-    char c2 = input[2] & 0x3F;
-    char c3 = input[3] & 0x3F;
-
-    int unicodePoint = ((c & 0x07) << 18) | (c1 << 12) | (c2 << 6) | c3;
-    unicodePoint -= 0x10000;
-
-    int surrogates[2] { 0xD800 | (unicodePoint >> 10), 0xDC00 | (unicodePoint & 0x3FF) };
-
-    output[0] = (surrogates[0]>>12 & 0x0F) | 0xE0;
-    output[1] = (surrogates[0]>>6  & 0x3F) | 0x80;
-    output[2] = (surrogates[0]     & 0x3F) | 0x80;
-    output[3] = (surrogates[1]>>12 & 0x0F) | 0xE0;
-    output[4] = (surrogates[1]>>6  & 0x3F) | 0x80;
-    output[5] = (surrogates[1]     & 0x3F) | 0x80;
-}
-
-ssize_t litecore::jni::UTF8ToModifiedUTF8(const char* input, const char** output, size_t len){
-    // https://github.com/android-ndk/ndk/issues/283
-    size_t extraBytes = 0;
-    const auto unsignedInput = (const uint8_t *)input;
-
-    // Need to figure out the actual length since each
-    // 4-bytes of real UTF-8 is 6 bytes of modified UTF-8
-    // but convert the bytes while we are at it
-    for(size_t i = 0; i < len; i++) {
-        if(unsignedInput[i] >= 0xF0) {
-            // 0xF0 and above marks 4-byte sequences
-            extraBytes += 2;
-
-            // 3 + 1 from next loop iteration = 4 bytes advance
-            i += 3;
-        }
-    }
+namespace litecore {
+    namespace jni {
+        std::string JstringToUTF8(JNIEnv *env, jstring jstr);
 
-    if(extraBytes == 0) {
-        // No modifications necessary
-        *output = nullptr;
-        return len;
+        jstring UTF8ToJstring(JNIEnv *env, std::string str);
     }
+}
 
-    size_t newStrLen = len + extraBytes;
-    char* newBytes = (char *)malloc(newStrLen);
-    if(newBytes == nullptr) {
-        *output = nullptr;
-        return -1;
+// Java uses Modified-UTF-8, not UTF-8: Attempting to decode a real UTF-8 string will cause a failure that looks like:
+//   art/runtime/check_jni.cc:65] JNI DETECTED ERROR IN APPLICATION: input is not valid Modified UTF-8: illegal start byte ...
+//   art/runtime/check_jni.cc:65]     string: ...
+//   art/runtime/check_jni.cc:65]     in call to NewStringUTF
+// See:
+//   https://stackoverflow.com/questions/35519823/jni-detected-error-in-application-input-is-not-valid-modified-utf-8-illegal-st
+// The strategy here is to use standard C functions to convert the UTF-8 directly to UTF-16, which Java handles nicely.
+// The following two functions are taken from this repo:
+//   https://github.com/incanus/android-jni/blob/master/app/src/main/jni/JNI.cpp#L57-L86
+
+jstring litecore::jni::UTF8ToJstring(JNIEnv *env, std::string str) { // < !!! SHOULD BE REF?
+    std::u16string ustr = std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>().from_bytes(str);
+
+    jstring jstr = env->NewString(reinterpret_cast<const jchar *>(ustr.c_str()), ustr.size());
+    if (jstr == nullptr) {
+        env->ExceptionDescribe();
```

Comment:
This will log to STDERR which doesn't help anything. Is it better to throw exception?